cpp/src/arrow/array/concatenate.cc

```diff
@@ -818,8 +937,31 @@ Result<std::shared_ptr<Array>> Concatenate(const ArrayVector& arrays, MemoryPool
   }
 
   std::shared_ptr<ArrayData> out_data;
-  RETURN_NOT_OK(ConcatenateImpl(data, pool).Concatenate(&out_data));
+  ErrorHints hints;
+  auto status = ConcatenateImpl(data, pool).Concatenate(&out_data, &hints);
+  if (!status.ok()) {
+    if (hints.suggested_cast) {
+      DCHECK(status.IsInvalid());
+      *out_suggested_cast = std::move(hints.suggested_cast);
```

Comment:
```
/// \param[out] out_suggested_cast if non-NULL, the function might set it to a cast
///                                suggestion that would allow concatenating the arrays
///                                without overflow of offsets (e.g. string to
///                                large_string)
```

So `out_suggested_cast` might be nullptr isn't it?